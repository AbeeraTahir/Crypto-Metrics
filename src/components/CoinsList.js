import React, { useEffect, useState } from 'react';
import { Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { getCoins } from '../redux/AllCoins/coinSlice';

const CoinsList = () => {
  const dispatch = useDispatch();
  const coins = useSelector((state) => state.coins.coinsData);
  const [search, setSearch] = useState('');

  useEffect(() => {
    if (coins.length === 0) {
      dispatch(getCoins());
    }
  }, [dispatch, coins.length]);

  const handleChange = (e) => {
    setSearch(e.target.value);
  };

  const searchedCoin = coins.filter((coin) => coin.name.toLowerCase()
    .match(search.toLowerCase()));

  return (
    <div>
      <input
        type="search"
        placeholder="Search here"
        onChange={handleChange}
        value={search}
      />
      <div>
        {searchedCoin.map((coin) => (
          <Link to={`/details/${coin.id}`} key={coin.id}>
            <div>
              <img src={coin.icon} alt="coin icon" />
              <h2>{coin.name}</h2>
              <p>
                {(coin.price / 1000).toFixed(1)}
                K
              </p>
            </div>
          </Link>
        ))}
      </div>
    </div>
  );
};

export default CoinsList;
