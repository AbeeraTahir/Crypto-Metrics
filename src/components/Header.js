import React from 'react';
import { Link } from 'react-router-dom';
import back from '../assets/arrow_back.svg';
import '../styles/Header.css';
import menu from '../assets/menu.png';

const Header = () => (
  <header>
    <div className="left">
      <Link to="/">
        <img src={back} alt="back" />
      </Link>
      <h1>Cryptocurrency</h1>
    </div>
    <img src={menu} alt="menu" />
  </header>
);

export default Header;
