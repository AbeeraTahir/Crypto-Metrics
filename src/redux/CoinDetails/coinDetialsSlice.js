import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import fetchCoinDetails from './fetchCoinDetailsAPI';

export const getCoinDetails = createAsyncThunk('details/fetchCoinDetails', async (id) => {
  const coinDetails = await fetchCoinDetails(id);
  const { coin } = coinDetails;
  return coin;
});

const initialState = { loading: false, coinDetails: [], error: '' };

const coinDetailsSlice = createSlice({
  name: 'details',
  initialState,
  extraReducers: (builder) => {
    builder.addCase(getCoinDetails.pending, (state) => {
      const newState = { ...state, loading: true };
      return newState;
    });
    builder.addCase(getCoinDetails.fulfilled, (state, action) => {
      const newState = { ...state, coinDetails: action.payload, loading: false };
      return newState;
    });
    builder.addCase(getCoinDetails.rejected, (state, action) => {
      const newState = { ...state, coinDetails: [], error: action.error.message };
      return newState;
    });
  },
});

export default coinDetailsSlice.reducer;
