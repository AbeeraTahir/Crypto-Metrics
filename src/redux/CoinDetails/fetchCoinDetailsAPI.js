const fetchCoinDetails = async (id) => {
  const response = await fetch(`https://api.coinstats.app/public/v1/coins/${id}`);
  const responseJSON = await response.json();
  return responseJSON;
};

export default fetchCoinDetails;
