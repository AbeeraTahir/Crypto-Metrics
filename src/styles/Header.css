header {
  display: flex;
  padding: 0.8em 0.4em;
  justify-content: space-between;
  align-items: center;
  background-color: var(--light-blue);
}

header img {
  width: 25px;
  height: 25px;
  margin-right: 0.4em;
  margin-top: 0.2em;
}

.left {
  display: flex;
  align-items: center;
  gap: 0.2em;
}

.left h1 {
  font-size: 1.3rem;
  font-weight: 800;
  color: var(--text-color);
}

.left img {
  width: 30px;
  height: 30px;
  margin-top: 5px;
}
