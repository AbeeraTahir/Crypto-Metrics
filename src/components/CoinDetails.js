import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';
import { getCoinDetails } from '../redux/CoinDetails/coinDetialsSlice';
import '../styles/coinDetails.css';

const CoinDetailsPage = () => {
  const dispatch = useDispatch();
  const params = useParams();
  const coinDetails = useSelector((state) => state.details.coinDetails);

  useEffect(() => {
    dispatch(getCoinDetails(params.id));
  }, [dispatch, params.id]);

  return (
    <div className="coin-details">
      <div className="hero">
        <img src={coinDetails.icon} alt="coin-icon" />
        <div className="hero-head">
          <h2>
            {coinDetails.name}
          </h2>
        </div>
      </div>
      <div className="details">
        <h2>
          {coinDetails.name}
          {' '}
          Details
        </h2>
        <ul>
          <li>
            <span>Name:</span>
            <span>{coinDetails.name}</span>
          </li>
          <li>
            <span>Symbol:</span>
            <span>{coinDetails.symbol}</span>
          </li>
          <li>
            <span>Rank:</span>
            <span>{coinDetails.rank}</span>
          </li>
          <li>
            <span>Price to USD:</span>
            <span>
              {' ~$'}
              {coinDetails.price < 1000
                ? coinDetails.price.toFixed(2) : (coinDetails.price / 1000).toFixed(1)}
              {coinDetails.price > 1000 ? 'K' : ''}
            </span>
          </li>
          <li>
            <span>Daily Change:</span>
            <span>
              {coinDetails.priceChange1d}
              %
            </span>
          </li>
          <li>
            <span>Market Cap:</span>
            <span>
              {(coinDetails.marketCap / 1000000000).toFixed(1)}
              B
            </span>
          </li>
        </ul>
      </div>
    </div>
  );
};

export default CoinDetailsPage;
