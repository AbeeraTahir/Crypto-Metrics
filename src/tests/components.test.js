import React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import store from '../redux/configureStore';
import Header from '../components/Header';
import CoinsList from '../components/CoinsList';
import CoinDetails from '../components/CoinDetails';

describe('Tests for react components', () => {
  it('Header component should render correctly', () => {
    const header = render(
      <React.StrictMode>
        <Router>
          <Provider store={store}>
            <Header />
          </Provider>
        </Router>
      </React.StrictMode>,
    );
    expect(header).toMatchSnapshot();
  });
  it('Coin list component should render correctly', () => {
    const list = render(
      <React.StrictMode>
        <Router>
          <Provider store={store}>
            <CoinsList />
          </Provider>
        </Router>
      </React.StrictMode>,
    );
    expect(list).toMatchSnapshot();
  });
  it('Coin details component should render correctly', () => {
    const details = render(
      <React.StrictMode>
        <Router>
          <Provider store={store}>
            <CoinDetails />
          </Provider>
        </Router>
      </React.StrictMode>,
    );
    expect(details).toMatchSnapshot();
  });
});
