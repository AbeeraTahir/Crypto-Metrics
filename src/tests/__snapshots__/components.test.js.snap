// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for react components Coin details component should render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="coin-details"
      >
        <div
          class="hero"
        >
          <img
            alt="coin-icon"
          />
          <div
            class="hero-head"
          >
            <h2 />
          </div>
        </div>
        <div
          class="details"
        >
          <h2>
             
            Details
          </h2>
          <ul>
            <li>
              <span>
                Name:
              </span>
              <span />
            </li>
            <li>
              <span>
                Symbol:
              </span>
              <span />
            </li>
            <li>
              <span>
                Rank:
              </span>
              <span />
            </li>
            <li>
              <span>
                Price to USD:
              </span>
              <span>
                 ~$
                NaN
              </span>
            </li>
            <li>
              <span>
                Daily Change:
              </span>
              <span>
                %
              </span>
            </li>
            <li>
              <span>
                Market Cap:
              </span>
              <span>
                NaN
                B
              </span>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="coin-details"
    >
      <div
        class="hero"
      >
        <img
          alt="coin-icon"
        />
        <div
          class="hero-head"
        >
          <h2 />
        </div>
      </div>
      <div
        class="details"
      >
        <h2>
           
          Details
        </h2>
        <ul>
          <li>
            <span>
              Name:
            </span>
            <span />
          </li>
          <li>
            <span>
              Symbol:
            </span>
            <span />
          </li>
          <li>
            <span>
              Rank:
            </span>
            <span />
          </li>
          <li>
            <span>
              Price to USD:
            </span>
            <span>
               ~$
              NaN
            </span>
          </li>
          <li>
            <span>
              Daily Change:
            </span>
            <span>
              %
            </span>
          </li>
          <li>
            <span>
              Market Cap:
            </span>
            <span>
              NaN
              B
            </span>
          </li>
        </ul>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Tests for react components Coin list component should render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="coins-container"
      >
        <img
          alt="banner"
          src="banner.jpg"
        />
        <div
          class="search-field"
        >
          <input
            placeholder="Search cryptocurrency"
            type="search"
            value=""
          />
          <img
            alt="search icon"
            src="find.svg"
          />
        </div>
        <div
          class="coins-list"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="coins-container"
    >
      <img
        alt="banner"
        src="banner.jpg"
      />
      <div
        class="search-field"
      >
        <input
          placeholder="Search cryptocurrency"
          type="search"
          value=""
        />
        <img
          alt="search icon"
          src="find.svg"
        />
      </div>
      <div
        class="coins-list"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Tests for react components Header component should render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header>
        <div
          class="left"
        >
          <a
            href="/"
          >
            <img
              alt="back"
              src="arrow-back-ios.svg"
            />
          </a>
          <h1>
            Cryptocurrency
          </h1>
        </div>
        <div
          class="right"
        >
          <img
            alt="mic"
            src="mic.svg"
          />
          <img
            alt="settings"
            src="settings.svg"
          />
        </div>
      </header>
    </div>
  </body>,
  "container": <div>
    <header>
      <div
        class="left"
      >
        <a
          href="/"
        >
          <img
            alt="back"
            src="arrow-back-ios.svg"
          />
        </a>
        <h1>
          Cryptocurrency
        </h1>
      </div>
      <div
        class="right"
      >
        <img
          alt="mic"
          src="mic.svg"
        />
        <img
          alt="settings"
          src="settings.svg"
        />
      </div>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
